File: src/CourseSystem.Persistence/Courses/Course.cs
using CourseSystem.Persistence.Abstractions;
using CourseSystem.Persistence.Categories;
using CourseSystem.Persistence.Languages;

namespace CourseSystem.Persistence.Courses;

public class Course : BaseEntity
{
    public string Name { get; private set; }

    public string? ShortDescription { get; private set; }

    public string? Description { get; private set; }

    public int? CategoryId { get; private set; }

    public int? LanguageId { get; private set; }

    public int? QuestionAnswerCount { get; private set; }

    public bool IsActive { get; private set; }

    public string? Slug { get; private set; }

    public virtual Category? Category { get; private set; }

    public virtual Language? Language { get; private set; }

    protected Course()
    {
    }

    private Course(string name, string? shortDescription, string? description, int? categoryId, int? languageId,
        int? questionAnswerCount, bool isActive, string? slug)
    {
        Name = name;
        ShortDescription = shortDescription;
        Description = description;
        CategoryId = categoryId;
        LanguageId = languageId;
        QuestionAnswerCount = questionAnswerCount;
        IsActive = isActive;
        Slug = slug;
    }

    public static Course Create(string name, string? shortDescription, string? description, int? categoryId,
        int? languageId, int? questionAnswerCount, bool isActive, string? slug)
    {
        return new Course(name, shortDescription, description, categoryId, languageId, questionAnswerCount, isActive,
            slug);
    }

    public void UpdateCourse(string name, string? shortDescription, string? description, int? categoryId, int? languageId,
        int? questionAnswerCount, bool isActive, string? slug)
    {
        Name = name;
        ShortDescription = shortDescription;
        Description = description;
        CategoryId = categoryId;
        LanguageId = languageId;
        QuestionAnswerCount = questionAnswerCount;
        IsActive = isActive;
        Slug = slug;
    }
}

File: course-frontend/src/types/Course.ts
// ===== GET /api/courses (Get All)
export interface GetAllCoursesResponse {
  courses: GetCourseResponse[];
}

// ===== GET /api/courses/{id}
export interface GetCourseResponse {
  id: number;
  name: string;
  shortDescription: string;
  description: string;
  categoryId?: number;
  categoryName?: string;
  languageId?: number;
  languageCode?: string;
  questionAnswerCount?: number;
  isActive: boolean;
  slug?: string;
  createdAt: string;
  updatedAt: string;
}

// ===== Represents a course inside lists (you may reuse GetCourseResponse too)
export interface Course {
  id: number;
  name: string;
  shortDescription?: string;
  description?: string;
  
  categoryId?: number | null;    // <--- FIX
  categoryName?: string | null;

  languageId?: number | null;     // <--- FIX
  languageCode?: string | null;   // <--- returned only

  questionAnswerCount?: number;
  isActive: boolean;
  slug?: string;
  createdAt: string;
  updatedAt: string;
}

// ===== POST /api/courses (Create)
export interface CreateCourseRequest {
  name: string;
  shortDescription?: string;
  description?: string;
  categoryId: number;
  languageId: number;
  questionAnswerCount?: number;
  isActive: boolean;
  slug?: string;
}

export interface CreateCourseResponse {
  id: number;
}

// ===== PUT /api/courses/{id} (Update)
export interface UpdateCourseRequest {
  courseId: number;
  name: string;
  shortDescription?: string;
  description?: string;
  categoryId: number;
  languageId: number;
  questionAnswerCount?: number;
  isActive: boolean;
  slug?: string;
}

export interface UpdateCourseResponse {
  id: number;
}

// ===== DELETE /api/courses/{id}
export interface DeleteCourseResponse {
  id: number;
}
